core/src/main/java/org/apache/calcite/rel/rules/ReduceExpressionsRule.java

```diff
@@ -164,6 +164,10 @@ public FilterReduceExpressionsRule(Class<? extends Filter> filterClass,
           "ReduceExpressionsRule(Filter)");
     }
 
+    @Override public boolean matches(RelOptRuleCall call) {
+      return RelOptUtil.getVariablesUsed(call.rel(0).getInputs()).isEmpty();
```

Comment:
I am not sure about this change. I've seen the justification that without this extra check the planner may generate a non-optimal plans for a given query but this may mean that we are just missing some additional rules. Moreover, this change breaks backward compatibility since it limits the applicability of the rule without good reason.

Reply:
removed